setup.py

```diff
@@ -5,28 +5,53 @@
 from setuptools.command.install import install
 from setuptools.command.develop import develop
 
-VERSION = ""
+
+def read(fname):
+    try:
+        with open(os.path.join(os.path.dirname(__file__), fname)) as f:
+            data = f.read()
+            return data
+    except BaseException:
+        return None
+
+
+def write(fname, data):
+    try:
+        with open(os.path.join(os.path.dirname(__file__), fname), 'w') as f:
+            f.write(data)
+    except BaseException:
+        pass
+
+
+TRAVIS_BUILD_NUMBER = os.environ.get("TRAVIS_BUILD_NUMBER", "0")
+TRAVIS_TAG = os.environ.get("TRAVIS_TAG", None)
+
+VERSION = read('studio/.version')
```

Comment:
Maybe I am missing some obvious best practice, but how will the tag propagate into the python version (so that you can do `studio.__version__` in the python code) if we are distributing it as source?

Reply:
In fact, the community solutions that I have found are using git tags, but when in PyPI package, still rely on the file (for instance http://blogs.nopcode.org/brainstorm/2013/05/20/pragmatic-python-versioning-via-setuptools-and-git-tags/), 
which does not seem to be fundamentally different from what I am doing here

Reply:
Could the following help to avoid using hand rolling of files ?
https://pypi.python.org/pypi/setuptools_scm_git_archive
https://pypi.python.org/pypi/setuptools_scm

Reply:
Let me take a look

Reply:
okay, so I switched to use setuptools_scm. 
The bottom line is this - when it's a tagged release, version is tag; when it's not, the version is tag.post<distance>
where distance is being computed by the setuptools_scm magic (I suspect it's the shortest distance in git DAG). And all those are being pushed into pypi when built successfully on master.